### ISSUE TITLE
Merged sce objects

### ISSUE THREAD

**Issue by bioguy2018** [HelenaLC/CATALYST#231]
Dear Helena,
I have merged two sce objects of my analysis using cbind(). Everything is okay but when I need to draw a heatmap I face a problem as the heatmap function looks into the cluster_id and since my datasets are merged I have stored it in merged_sce[[Cluster]].
Is there a way to overcome this issue?

Thank you so much!

**Comment by HelenaLC** [comment]
I'm sorry, but I can't reproduce the issue from this information. Can you clarify what you mean by "I have stored it in merged_sce[[Cluster]]"? What have you stored there? And what is `Cluster`? Is that a variable or supposed to be a character, i.e. `merged_sce[["Cluster"]]`? Ideally, could you please post the code leading up to this issue? Even more ideally, using the example data from the package...

**Comment by bioguy2018** [comment]
Dear Helena,
Please excuse my minimal provided information I will try to explain it better.

```
# I have two datasets 
sce1
sce2
sce1= cluster(sce1)
sce2= cluster(sce2)
# I decide for 5 clusters for sce1 and 4 for sce2 and I rename them using
levels(sce1@metadata[["cluster_codes"]][["meta5"]])=c("1_a","1_b","1_c","1_d","1_e")
levels(sce2@metadata[["cluster_codes"]][["meta4"]])=c("2_a","2_b","2_c","2_d")
# I store these clusters as follow in colData
sce1[["Cluster"]] <- cluster_ids(sce1, "meta5")
sce2[["Cluster"]] <- cluster_ids(sce2, "meta4")
# I merge the datasets
merged_sce=cbind(sce1,sce2)
# now I want to plot a heatmap
plotExprHeatmap(merged_sce, features = "type", by = "Cluster", scale = "last", q = 0.01, perc = TRUE, col_dend = F,row_clust = T,bars = T,col_clust = F) 
```
The last piece of code is not possible as the function either looks into sample_id or cluster_id and since during the merge the meta clusters are not merged and are kept separately (I just get a duplicate from one of them) it is not possible for me to use cluster_id. That's the reason I stored my clusters separately in [[Cluster]] because then in merged_sce[["Cluster"]] they are merged and I can access all the cells using the corresponding cluster.
------------------
ps. I need to analyze these data separately due to batch effect and this would cause a huge problem when the data is together transformed and analyzed but since my purpose is to find overlapping clusters at the end. I decided to merge them after an initial separate analysis and using the heatmap function I aggregate and scale over the values.

**Comment by HelenaLC** [comment]
Yes, so this is a little more complicated. `CATALYST` uses the `cluster_id` cell metadata (`colData`) column to look up the cluster IDs in the `metadata()$cluster_codes` field. So, both the `cluster_id`s & `cluster_codes` need to be made unique across samples. Below, I wrote a minimal example of how to do this (should should be able to run it as is). Basically, I am pre-pending a sample-identified to all cluster IDs (`"one"` for the 1st SCE, and `"two"` for the 2nd SCE). My apologies that this is so complicated. Please let me know if you can follow along or need additional help! In the long run, I think I can try including a wrapper function to do this for an arbitrary number of samples. Putting it on my list of feature requests!

```
library(CATALYST)
data(PBMC_fs, PBMC_panel, PBMC_md)
sce <- prepData(PBMC_fs, PBMC_panel, PBMC_md)

# make up 2 SCEs by clustering with different seeds
set.seed(1); a <- cluster(sce, verbose = FALSE)
set.seed(2); b <- cluster(sce, verbose = FALSE)

# make cluster IDs uniquely identifyable
l <- list(one = a, two = b)
l <- lapply(names(l), \(i) {
    sce <- l[[i]]
    ids <- sce$cluster_id
    ids <- paste(i, ids, sep = ".")
    sce$cluster_id <- factor(ids)
    codes <- cluster_codes(sce)
    codes <- lapply(codes, \(ids) paste(i, ids, sep = "."))
    codes <- data.frame(codes, stringsAsFactors = TRUE)
    metadata(sce)$cluster_codes <- codes
    return(sce)
})

# join data & cluster codes
c <- do.call(cbind, l)
codes <- lapply(l, cluster_codes)
codes <- do.call(rbind, codes)
codes <- data.frame(codes)
metadata(c)$cluster_codes <- codes

plotExprHeatmap(c, 
    features = "type", 
    by = "cluster_id", 
    k = "meta10")
```

**Comment by bioguy2018** [comment]
Dear Helena,
Thank you so much for your answer and efforts! Sadly I couldn't reproduce your code.
I am not sure if \(i) and \(ids) in the second part of the code should be there or how should I implement it.

**Comment by HelenaLC** [comment]
By "couldn't reproduce" do you mean you couldn't run the exact code above? Or couldn't apply it to your data? I ran it in a fresh R session and it works for me.

Here, `i` is the name of the sample in a list of SCEs, so in your case, you could do something like `l = list(a = sce1, b = sce2)` or anything. The point is, that I am appending a unique identifier to each SCE. You could also use
```
l = list(sce1, sce2)
lapply(seq_along(l), \(i) ...)
```
which would pre-pend a 1/2 before the cluster IDs, but I find this more difficult to interpret, as the IDs will look like 1.1 1.2 1.3 ... 2.1 2.2 2.3 ...

And, `ids` are the cluster IDs stored inside of the `colData`. These are required to match with the ones in `cluster_codes`. So I am pre-pending `i` to both, the `cluster_id` column, and each column in `cluster_codes`.

**Comment by bioguy2018** [comment]
Dear Helena,
I couldn't run it due to syntax errors.
`Error: unexpected input in "l <- lapply(names(l), \"`
I got the concept I guess and yesterday I tried to do it in a different way using below code

```
data(PBMC_fs, PBMC_panel, PBMC_md)
sce <- prepData(PBMC_fs, PBMC_panel, PBMC_md)
set.seed(1); a <- cluster(sce, verbose = FALSE)
set.seed(2); b <- cluster(sce, verbose = FALSE)
#making unique names
names=colnames(a@metadata[["cluster_codes"]])
i=2
while(i <= length(names)) {
    levels(a@metadata[["cluster_codes"]][[names[i]]])=paste("1",levels(a@metadata[["cluster_codes"]][[names[i]]]),sep="_")
    i=i+1
}
i=1
while(i <= length(levels(a@colData$cluster_id))) {
  levels(a@colData$cluster_id)[i]=paste("1",levels(a@colData$cluster_id)[i],sep="_")
  i=i+1
}

names=colnames(b@metadata[["cluster_codes"]])
i=2
while(i <= length(names)) {
  levels(b@metadata[["cluster_codes"]][[names[i]]])=paste("2",levels(b@metadata[["cluster_codes"]][[names[i]]]),sep="_")
  i=i+1
}
i=1
while(i <= length(levels(a@colData$cluster_id))) {
  levels(b@colData$cluster_id)[i]=paste("2",levels(b@colData$cluster_id)[i],sep="_")
  i=i+1
}

l <- list(one = a, two = b)
c <- do.call(cbind, l)
plotExprHeatmap(c, 
                features = "type", 
                by = "cluster_id", 
                k = "meta2")

```
Everything works good but upon merging and plotting I get:
`Error: Invalid depth
Run `rlang::last_error()` to see where the error occurred.
> rlang::last_error()
<error/rlang_error>
Invalid depth
Backtrace:
 1. CATALYST::plotExprHeatmap(...)
 2. CATALYST:::.do_agg()
 3. CATALYST:::.agg(x, by, fun, assay)
 4. purrr::map_depth(...)
 5. purrr:::map_depth_rec(...)
Run `rlang::last_trace()` to see the full context`

**Comment by HelenaLC** [comment]
Aha, okay, looks like you are not using R 4.1, in which case you simply need to replace `\(...)` with `function(...)` & it should run.

**Comment by bioguy2018** [comment]
Dear Helena,
Perfect! I suspected that it might be due to a different R version but when I was googling I couldn't find the syntax / in my skimming but thank you so much!!
I will try it using my own data and if I needed any further help I will bother you. If you got time I would appreciate it if you let me know why the solution I tried to run led into "Invalid depth".  I would very much appreciate it ! This will help me to think better for future problems
